waku/v2/protocol/waku_store/protocol.nim

```diff
@@ -213,19 +225,26 @@ proc handleMessage*(w: WakuStore, pubsubTopic: string, msg: WakuMessage) =
     # The message is ephemeral, should not be stored
     return
   
+  if not isValidMessage(msg):
+    waku_store_errors.inc(labelValues = [invalidMessage])
+    return
+
+
   let insertStartTime = getTime().toUnixFloat()
-    
-  let now = getNanosecondTime(getTime().toUnixFloat())
-  let index = Index.compute(msg, receivedTime=now, pubsubTopic=pubsubTopic)
+  
+  block:
+    let
+      msgDigest = computeDigest(msg) 
+      msgReceivedTime = if msg.timestamp > 0: msg.timestamp
+                        else: getNanosecondTime(getTime().toUnixFloat()) 
 
-  trace "handling message", topic=pubsubTopic, index=index
+    trace "handling message", pubsubTopic=pubsubTopic, contentTopic=msg.contentTopic, timestamp=msg.timestamp, digest=msgDigest
 
-  # Add messages to persistent store, if present
-  let putStoreRes = w.store.put(index, msg, pubsubTopic)
-  if putStoreRes.isErr():
-    debug "failed to insert message to persistent store", index=index, err=putStoreRes.error
-    waku_store_errors.inc(labelValues = [insertFailure])
-    return
+    let putStoreRes = w.store.put(pubsubTopic, msg, msgDigest, msgReceivedTime)
```

Comment:
This is the logic that sets the message store timestamp:

* If the message timestamp (a.k.a. `senderTimestamp`) is present, use that timestamp as the store timestamp.
* If the message timestamp is equal to 0 (which means either the timestamp is not present or was set to 0), we insert the message with a store timestamp equal to the arrival time.

Note that at this point, we have already checked if the message is valid.